src/source/shift.rs

```diff
@@ -4,141 +4,136 @@ use json;
+        })
     }
 
-    pub fn with_path<P: AsRef<Path>>(bin_path: P) -> Self {
-        Shift {
-            bin_path: bin_path.as_ref().to_path_buf()
-        }
+    pub fn create_cmd(&self, name: &str) -> Command {
+        let mut cmd = Command::new(&self.bin_path);
+        cmd.env("NODE_PATH", "node_modules");
+        cmd.arg(&self.memory);
+        cmd.arg(self.scripts_path.join(name));
+        cmd
     }
+}
 
-    fn parse_script_output(&self, script: &str) -> Result<String, Error> {
-        debug!(target: "Shift", "Preparing script {}", script);
-
-        let script = format!(r##"
-        var result;
-        try {{
-            result = (function() {{
-                {}
-            }})();
-        }} catch (ex) {{
-            console.warn(ex);
-            /* rethrow */ throw ex;
-        }};
-        var process = require('process');
-        /* See crates/binjs_io/src/escaped_wtf8.rs */
-        result = result
-            .replace(/[\u007F\uD800-\uDFFF]/ug, function(m) {{
-                if (m == "\u007F") {{
-                    return "\u007F007F";
-                }}
-                return "\u007F" + m.charCodeAt(0).toString(16);
-            }});
-        process.stdout.write(result);
-        console.warn(result);
-        "##,
-        script);
-
-        debug!(target: "Shift", "Launching script {}", script);
-
-        let node_memory = match env::var("NODE_MAX_OLD_SPACE_SIZE") {
-            Err(_) => String::from("--max_old_space_size=2048"),
-            Ok(v) => format!("--max_old_space_size={}", v)
-        };
+struct ScriptIO {
+    input: LineWriter<ChildStdin>,
+    output: Lines<BufReader<ChildStdout>>,
+}
 
-        let mut child = Command::new(&*self.bin_path)
-            .arg(node_memory)
-            .env("NODE_PATH", "node_modules")
+pub struct Script(Mutex<ScriptIO>);
+
+impl Script {
+    pub fn try_new(node_config: &NodeConfig, name: &str) -> Result<Self, Error> {
+        let child = node_config.create_cmd(name)
             .stdin(Stdio::piped())
             .stdout(Stdio::piped())
-            .stderr(Stdio::piped())
+            .stderr(Stdio::inherit())
             .spawn()
             .map_err(Error::CouldNotLaunch)?;
 
-        if let Some(ref mut stdin) = child.stdin {
-            stdin.write(script.as_bytes())
-                .map_err(Error::ExecutionError)?;
-        }
-
-        let output = child.wait_with_output()
-            .map_err(Error::ExecutionError)?;
+        Ok(Script(Mutex::new(ScriptIO {
+            input: LineWriter::new(child.stdin.unwrap()),
+            output: BufReader::new(child.stdout.unwrap()).lines(),
+        })))
+    }
 
-        debug!(target: "Shift", "Output {:?}", output);
-        if !output.status.success() {
-            error!(target: "Shift", "Script error {}", String::from_utf8(output.stderr).unwrap());
-            return Err(Error::ReturnedError(output.status));
+    pub fn transform(&self, input: &JSON) -> Result<JSON, Error> {
```

Comment:
Transform what into what?

Again, documentation would be useful :)

Reply:
Anything into anything (as all scritpts are different, this is just a generic API to pass some input to them and get some output back).